File: cmd/2015/day11/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func getNextPassword(passwordChars []rune) []rune {
	for i := len(passwordChars) - 1; i >= 0; i-- {
		if passwordChars[i] == 'z' {
			passwordChars[i] = 'a'
		} else {
			passwordChars[i]++
			break
		}
	}
	return passwordChars
}

func isValidPassword(passwordChars []rune) bool {
	for _, char := range passwordChars {
		if char == 'i' || char == 'o' || char == 'l' {
			return false
		}
	}

	hasStraight := false
	for i := range len(passwordChars) - 2 {
		if passwordChars[i]+1 == passwordChars[i+1] &&
			passwordChars[i+1]+1 == passwordChars[i+2] {
			hasStraight = true
			break
		}
	}
	if !hasStraight {
		return false
	}

	pairs := make(map[rune]bool)
	for i := 0; i < len(passwordChars)-1; i++ {
		if passwordChars[i] == passwordChars[i+1] {
			pairs[passwordChars[i]] = true
			i++
		}
	}

	return len(pairs) >= 2
}

func part1(password string) string {
	passwordChars := []rune(password)
	for !isValidPassword(passwordChars) {
		passwordChars = getNextPassword(passwordChars)
	}
	return string(passwordChars)
}

func part2(password string) string {
	password = part1(password)
	passwordChars := []rune(password)
	for {
		passwordChars = getNextPassword(passwordChars)
		if isValidPassword(passwordChars) {
			break
		}
	}
	return string(passwordChars)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	input := scanner.Text()
	fmt.Println("Part 1:", part1(input))
	fmt.Println("Part 2:", part2(input))
}

File: cmd/2015/day11/main.ts
function getNextPassword(passwordChars: string[]): string[] {
  for (let i = passwordChars.length - 1; i >= 0; i--) {
    if (passwordChars[i] === "z") {
      passwordChars[i] = "a";
    } else {
      passwordChars[i] = String.fromCharCode(
        passwordChars[i].charCodeAt(0) + 1
      );
      break;
    }
  }
  return passwordChars;
}

function isValidPassword(passwordChars: string[]): boolean {
  if (passwordChars.some((char) => ["i", "o", "l"].includes(char))) {
    return false;
  }

  let hasStraight = false;
  for (let i = 0; i < passwordChars.length - 2; i++) {
    if (
      passwordChars[i].charCodeAt(0) + 1 ===
        passwordChars[i + 1].charCodeAt(0) &&
      passwordChars[i + 1].charCodeAt(0) + 1 ===
        passwordChars[i + 2].charCodeAt(0)
    ) {
      hasStraight = true;
      break;
    }
  }
  if (!hasStraight) return false;

  const pairs = new Set<string>();
  for (let i = 0; i < passwordChars.length - 1; i++) {
    if (passwordChars[i] === passwordChars[i + 1]) {
      pairs.add(passwordChars[i]);
      i++;
    }
  }

  return pairs.size >= 2;
}

function part1(password: string): string {
  let passwordChars = password.split("");
  while (!isValidPassword(passwordChars)) {
    passwordChars = getNextPassword(passwordChars);
  }
  return passwordChars.join("");
}

function part2(password: string): string {
  password = part1(password);
  let passwordChars = password.split("");
  do {
    passwordChars = getNextPassword(passwordChars);
  } while (!isValidPassword(passwordChars));
  return passwordChars.join("");
}

if (import.meta.main) {
  const input = (await Bun.stdin.text()).trim();
  console.log("Part 1:", part1(input));
  console.log("Part 2:", part2(input));
}
